src/ctia/shutdown.clj

```diff
@@ -0,0 +1,28 @@
+(ns ctia.shutdown
+  (:require [clojure.tools.logging :as log]))
+
+(defonce shutdown-hooks (ref {:registered? false :hooks {}}))
+
+(defn register-hook! [key f]
+  (dosync
+   (alter shutdown-hooks assoc-in [:hooks key] f)))
+
+(defn shutdown-ctia! []
+  (doseq [[name hook] (:hooks @shutdown-hooks)]
+    (try
+      (hook)
+      (catch Exception excep
+        (log/error excep (format "Shutdown hook %s failed" name)))))
```

Comment:
I didn't put this in a transaction because I don't know how safe it is to rely on the agent-send-off thread pool at shutdown time.  (Hooks would have side effects, so I'd want to use send-off).  So calling shutdown-ctia! isn't completely thread safe.  Two threads theoretically could call it at the same time and trigger the hooks concurrently.  Its not likely to happen in practice, however.  Shutdown hooks should be written defensively anyhow.